api/envoy/config/route/v3/route_components.proto

```diff
@@ -1085,6 +1085,25 @@ message RetryPolicy {
     google.protobuf.Duration max_interval = 2 [(validate.rules).duration = {gt {}}];
   }
 
+  message RateLimitedRetryBackOff {
+    // Specifies the rate limited retry reset headers (like ``Retry-After`` or
+    // ``X-RateLimit-Reset``) to match against the response. If a request is
+    // being retried and the response contains one of these headers the value
+    // of the header will be used as the retry back off interval, with jitter added.
+    //
+    // The value of the header must be either:
+    //  1. An interval in seconds as an integer, for example: ``120``.
+    //  2. A unix timestamp as an integer, for example: ``1595320702``. The timestamp
+    //     must be in the future relative to the current system time.
+    repeated HeaderMatcher reset_headers = 1;
```

Comment:
Another thing to at least document is what happens when multiple conflicting headers are present

Reply:
> Another thing to at least document is what happens when multiple conflicting headers are present

I've added this.

Reply:
> Should this have a validation annotation that requires at least one value? Or is there a default?

Yes, it should. Without at least one header this will do nothing. I've added that.

> I'm not sure HeaderMatcher is the right type. You could easily specify a HeaderMatcher that would only match headers whose values do not conform to the described values. For example, matching a header named retry-after with a value prefixed by x is a legal HeaderMatcher. I also think inverted matches aren't likely to be useful. I think maybe we just want repeated string reset_headers?

I completely agree. I had misgivings about using HeaderMatcher here, because the only field that it makes senses to use is `name`. Using `exact_match`, `range_match`, `regex_match`, `present_match`, prefix/suffix or invert makes no sense, because as you say, the user might make the header no longer match.

The reason I used HeaderMatcher is that this is the UX pattern we have in the retry policy. All the other header related fields use it, so users don't have to learn a new type.

I'm happy to change it if you think it makes more sense to use `repeated string reset_headers`?

> Should the interval value be in milliseconds instead of seconds? I think that matches most of our other header-related intervals (timeouts and such).

This is not something we (ie. Envoy) control. The fields Retry-After and X-RateLimit-Reset are standardized:

- https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After
- https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html
- https://tools.ietf.org/id/draft-polli-ratelimit-headers-00.html

But in the wild they are not always used consistently. I discuss this in the design doc (linked in the first PR comment).

> I think we should consider some way to make the semantics of these values unambiguous. The interpretation of large values can change over time as current time gets larger. One way would be to maintain a separate set of interval vs. fixed time headers (though this implies an extra x-envoy header as well).

I see where you're coming from, but I'm skeptical here because I've found that the two commonly used representations today are a) seconds as an int and b) unix timestamp as an int. That's why I propose supporting these two. If in the future we wanted to support a date in the HTTP date format (which is mentioned in the Retry-After RFC but does not seem to be used much) then we'd have to add yet another parameter + header for this.

Reply:
I think it makes more sense to use `repeated string reset_headers`, but defer to the api shepherds.

For the ambiguity between interval & timestamp we could split the headers into those that take intervals and those that take timestamps. For example, either by having separate header lists or introducing a message with header name and an enum that describes the expected type of value.

Reply:
Okay, how about something like this:

```
  message RateLimitedRetryBackOff {
    message ResetHeader {
      enum ValueFormat {
        INT_SECONDS = 0;
        INT_UNIX_TIMESTAMP = 1;
      }

      string name = 1 [
        (validate.rules).string = {min_bytes: 1 well_known_regex: HTTP_HEADER_NAME strict: false}
      ];

      ValueFormat format = 2;
    }

    repeated ResetHeader reset_headers = 1 [(validate.rules).repeated = {min_items: 1}];

    google.protobuf.Duration reset_max_interval = 2 [(validate.rules).duration = {gt {}}];
  }
```

This makes it pretty extensible for adding other formats in the future if we need to.

As for the header based API, in the interest of not adding a separate header for each format we could do something like this:

```
x-envoy-ratelimited-reset-headers: Retry-After:INT_SECONDS, X-RateLimit-RESET:INT_UNIX_TIMESTAMP
```

Reply:
@mattklein123 or @alyssawilk would you like to provide direction?

Reply:
That addresses my concerns. But we do need someone from @envoyproxy/api-shepherds to approve in Harvey's stead.